subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/IncrementalAnnotationProcessorType.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.processing;
+
+/**
+ * The different kinds of annotation processors that the incremental compiler knows how to handle.
+ *
+ * In order to be treated incrementally, annotation processors need to use the originating element
+ * API provided by {@link javax.annotation.processing.Filer#createSourceFile(CharSequence, javax.lang.model.element.Element...)}
+ * and {@link javax.annotation.processing.Filer#createClassFile(CharSequence, javax.lang.model.element.Element...)}.
+ * Depending on how this API is used by the processor, it can be categorized as either SIMPLE or AGGREGATING.
+ *
+ * Processors that want to use {@link javax.annotation.processing.Filer#createResource(javax.tools.JavaFileManager.Location, CharSequence, CharSequence, javax.lang.model.element.Element...)}
+ * or {@link javax.annotation.processing.Filer#getResource(javax.tools.JavaFileManager.Location, CharSequence, CharSequence)}
+ * and processors that don't provide originating elements are unsupported.
+ *
+ * All processor types are currently treated as non-incremental. This will improve over time as we add more
+ * capabilities to the compiler.
+ */
+public enum IncrementalAnnotationProcessorType {
+    /**
+     * A processor whose generated files have exactly one originating element.
+     */
+    SIMPLE(false),
```

Comment:
:o: I wonder if constant names like `ONE_TO_ONE` and `MANY_TO_ONE` would be more readable. Especially if we eventually add `MANY_TO_MANY`.

Reply:
many-to-many is the same as many-to-one from our perspective. And one-to-many is the same as one-to-one. Only the incoming cardinality counts. That's why i chose the term "aggregating". But I'm not happy with the term "simple". Any other ideas?

Reply:
I see... What about `SINGLE_ORIGIN` vs `MULTIPLE_ORIGIN`?

Reply:
I like it!